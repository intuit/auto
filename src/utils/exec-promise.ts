import { spawn } from 'child_process';

/**
 * Wraps up running a command into a single promise,
 * returning the stdout as a string if the command succeeds
 * and throwing if it does not.
 *
 * @param cmd the command as a string to pass in
 */
export default async function execPromise(cmd: string, args?: string[]) {
  return new Promise<string>((completed, reject) => {
    const child = spawn(cmd, args || [], {
      cwd: process.cwd(),
      env: process.env,
      shell: true
    });

    let allStdout = '';
    child.stdout.on('data', async data => {
      const stdout = data.toString();
      allStdout += stdout;
    });

    let allStderr = '';
    child.stderr.on('data', data => {
      const stderr = data.toString();
      allStderr += stderr;
    });

    // This usually occurs during dev-time, when you have the wrong command
    child.on('error', err => {
      reject(
        new Error(`Failed to run '${cmd}' - ${err.message} \n\n\n${allStderr}`)
      );
    });

    child.on('exit', code => {
      // No code, no errors
      if (!code) {
        // Tools can occasionally print to stderr but not fail, so print that just in case.
        if (allStderr.length) {
          console.log(allStderr);
        }
        // Resolve the string of the whole stdout
        completed(allStdout.trim());
      } else {
        // The command bailed for whatever reason, print a verbose error message
        // with the stdout underneath
        let appendedStdErr = '';
        appendedStdErr += allStdout.length ? `\n\n${allStdout}` : '';
        appendedStdErr += allStderr.length ? `\n\n${allStderr}` : '';

        reject(new Error(`Running command '${cmd}' failed${appendedStdErr}`));
      }
    });
  });
}
