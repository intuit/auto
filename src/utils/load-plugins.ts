import * as path from 'path';
import { AutoRelease } from '../main';
import ChromeWebStorePlugin from '../plugins/chrome';
import NPMPlugin from '../plugins/npm';
import { ILogger } from './logger';

export type IPluginConstructor = new (options?: any) => IPlugin;

export interface IPlugin {
  name: string;
  apply(auto: AutoRelease): void;
}

type SupportedPlugin = 'npm' | 'chrome';
const plugins = new Map<SupportedPlugin, IPluginConstructor>([
  ['npm', NPMPlugin],
  ['chrome', ChromeWebStorePlugin]
]);

function isSupported(key: SupportedPlugin | string): key is SupportedPlugin {
  return !!plugins.get(key as SupportedPlugin);
}

function tryRequire(tryPath: string) {
  try {
    return require(tryPath);
  } catch (error) {
    return null;
  }
}

export default function loadPlugin(
  [pluginPath, options]: [SupportedPlugin | string, any],
  logger: ILogger
): IPlugin | undefined {
  let plugin: IPluginConstructor | undefined;

  if (isSupported(pluginPath)) {
    plugin = plugins.get(pluginPath);
  } else {
    plugin = tryRequire(pluginPath);

    if (!plugin) {
      plugin = tryRequire(path.join(process.cwd(), pluginPath));
    }
  }

  if (!plugin) {
    logger.log.warn(`Could not find plugin: ${pluginPath}`);
    return;
  }

  return new plugin(options);
}
