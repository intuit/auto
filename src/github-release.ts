import GHub from '@octokit/rest';
import * as fs from 'fs';
import { ICommit } from 'gitlog';
import { inc, ReleaseType } from 'semver';
import { promisify } from 'util';

import { SyncHook } from 'tapable';
import { ChangelogTitles, Label, Labels } from './config';
import GitHub, { IGitHubOptions, IPRInfo } from './git';
import LogParse, { IExtendedCommit, normalizeCommits } from './log-parse';
import SEMVER, { calculateSemVerBump } from './semver';
import execPromise from './utils/exec-promise';
import { dummyLog, ILogger } from './utils/logger';
import { makeGitHubReleaseHooks } from './utils/make-hooks';
import postToSlack from './utils/slack';

export type VersionLabel =
  | SEMVER.major
  | SEMVER.minor
  | SEMVER.patch
  | 'skip-release'
  | 'release'
  | 'prerelease';

export interface IGitHubReleaseOptions {
  jira?: string;
  slack?: string;
  githubApi?: string;
  name?: string;
  email?: string;
  owner?: string;
  repo?: string;
  skipReleaseLabels: string[];
  onlyPublishWithReleaseLabel?: boolean;
  noVersionPrefix?: boolean;
  changelogTitles?: {
    [label: string]: string;
  };
  versionLabels?: Map<VersionLabel, string>;
  plugins?: (string | [string, any])[];
}

export const defaultLabels = new Map<VersionLabel, string>();
defaultLabels.set(SEMVER.major, 'major');
defaultLabels.set(SEMVER.minor, 'minor');
defaultLabels.set(SEMVER.patch, 'patch');
defaultLabels.set('skip-release', 'skip-release');
defaultLabels.set('release', 'release');
defaultLabels.set('prerelease', 'prerelease');

export const defaultLabelsDescriptions = new Map<string, string>();
defaultLabelsDescriptions.set(
  SEMVER.major,
  'Increment the major version when merged'
);
defaultLabelsDescriptions.set(
  SEMVER.minor,
  'Increment the minor version when merged'
);
defaultLabelsDescriptions.set(
  SEMVER.patch,
  'Increment the patch version when merged'
);
defaultLabelsDescriptions.set(
  'skip-release',
  'Preserve the current version when merged'
);
defaultLabelsDescriptions.set(
  'release',
  'Create a release when this pr is merged'
);
defaultLabelsDescriptions.set(
  'prerelease',
  'Create a pre-release version when merged'
);
defaultLabelsDescriptions.set(
  'internal',
  'Changes only affect the internal API'
);
defaultLabelsDescriptions.set(
  'documentation',
  'Changes only affect the documentation'
);

const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);

export interface IGitHubReleaseHooks {
  onCreateLogParse: SyncHook<[LogParse]>;
}

export default class GitHubRelease {
  public readonly releaseOptions: IGitHubReleaseOptions;
  public readonly hooks: IGitHubReleaseHooks;

  private readonly logger: ILogger;
  private readonly github: GitHub;
  private readonly changelogTitles: { [label: string]: string };
  private readonly githubApi: string;
  private readonly versionLabels: Map<VersionLabel, string>;

  constructor(
    options: Partial<IGitHubOptions>,
    releaseOptions: IGitHubReleaseOptions = {
      skipReleaseLabels: []
    },
    logger: ILogger = dummyLog()
  ) {
    this.hooks = makeGitHubReleaseHooks();
    this.versionLabels = releaseOptions.versionLabels || defaultLabels;
    this.logger = logger;
    this.releaseOptions = releaseOptions;
    this.githubApi = releaseOptions.githubApi || 'https://api.github.com';
    options.baseUrl = this.githubApi;
    this.changelogTitles = releaseOptions.changelogTitles || {};

    if (!options.owner || !options.repo || !options.token) {
      throw new Error('Must set owner, repo, and GitHub token.');
    }

    this.logger.verbose.info('Options contain repo information.');

    // So that --verbose can be used on public CIs
    const tokenlessArgs = {
      ...options,
      token: `[Token starting with ${options.token.substring(0, 4)}]`
    };

    this.logger.verbose.info('Initializing GitHub API with:\n', tokenlessArgs);
    this.github = new GitHub(
      {
        owner: options.owner,
        repo: options.repo,
        token: options.token,
        baseUrl: options.baseUrl
      },
      this.logger
    );
  }

  public async generateReleaseNotes(
    from: string,
    to = 'HEAD'
  ): Promise<string> {
    const allCommits = await this.getCommits(from, to);
    const allPrCommits = await Promise.all(
      allCommits
        .filter(commit => commit.pullRequest)
        .map(async commit =>
          this.github.getCommitsForPR(Number(commit.pullRequest!.number))
        )
    );
    const allPrCommitHashes = allPrCommits
      .filter(Boolean)
      .reduce(
        (all, pr) => [...all, ...pr.map(subCommit => subCommit.sha)],
        [] as string[]
      );

    const commits = allCommits
      .filter(
        commit =>
          !allPrCommitHashes.includes(commit.hash) &&
          !commit.subject.includes('[skip ci]')
      )
      .map(commit => {
        if (commit.pullRequest) {
          return commit;
        }

        commit.labels = ['pushToMaster'];
        return commit;
      });

    const project = await this.github.getProject();
    const logParser = new LogParse(this.logger, {
      owner: this.github.options.owner,
      repo: this.github.options.repo,
      baseUrl: project.html_url,
      jira: this.releaseOptions.jira,
      versionLabels: this.versionLabels,
      changelogTitles: {
        ...new ChangelogTitles(),
        ...this.changelogTitles
      }
    });
    this.hooks.onCreateLogParse.call(logParser);
    logParser.loadDefaultHooks();

    return logParser.generateReleaseNotes(commits);
  }

  public async addToChangelog(
    releaseNotes: string,
    lastRelease: string,
    currentVersion: string,
    message = 'Update CHANGELOG.md [skip ci]'
  ) {
    this.logger.verbose.info('Adding new changes to changelog.');

    let version;

    if (lastRelease.match(/\d+\.\d+\.\d+/)) {
      version = await this.calcNextVersion(lastRelease);
    } else {
      // lastRelease is a git sha. no releases have been made
      const bump = await this.getSemverBump(lastRelease);
      version = inc(currentVersion, bump as ReleaseType);
    }

    this.logger.verbose.info('Calculated next version to be:', version);

    const date = new Date().toDateString();
    const prefixed =
      this.releaseOptions.noVersionPrefix ||
      (version && version.startsWith('v'))
        ? version
        : `v${version}`;

    let newChangelog = `# ${prefixed} (${date})\n\n${releaseNotes}`;

    if (fs.existsSync('CHANGELOG.md')) {
      this.logger.verbose.info('Old changelog exists, prepending changes.');
      const oldChangelog = await readFile('CHANGELOG.md', 'utf8');
      newChangelog = `${newChangelog}\n\n---\n\n${oldChangelog}`;
    }

    await writeFile('CHANGELOG.md', newChangelog);
    this.logger.verbose.info('Wrote new changelog to filesystem.');

    await execPromise('git', ['add', 'CHANGELOG.md']);
    await execPromise('git', ['commit', '-m', `"${message}"`, '--no-verify']);
    this.logger.verbose.info('Commited new changelog.');
  }

  public async getCommits(
    from: string,
    to = 'HEAD'
  ): Promise<IExtendedCommit[]> {
    this.logger.verbose.info(`Getting commits from ${from} to ${to}`);

    const gitlog = await this.github.getGitLog(from, to);

    this.logger.veryVerbose.info('Got gitlog:\n', gitlog);

    const labeledCommits = await this.addLabelsToCommits(gitlog);
    const commits = await this.getPRForRebasedCommits(labeledCommits);

    this.logger.veryVerbose.info('Added labels to commits:\n', commits);

    await Promise.all(
      commits.map(async commit => {
        let resolvedAuthors = [];

        if (commit.pullRequest) {
          const prCommits = await this.github.getCommitsForPR(
            Number(commit.pullRequest.number)
          );

          if (!prCommits) {
            return;
          }

          resolvedAuthors = await Promise.all(
            prCommits.map(async prCommit => {
              if (prCommit && prCommit.author) {
                return this.github.getUserByUsername(prCommit.author.login);
              }
            })
          );
        } else if (commit.authorEmail) {
          const author = await this.github.getUserByEmail(commit.authorEmail);
          resolvedAuthors.push(author);
        }

        commit.authors = resolvedAuthors.map(author => ({
          ...author,
          username: author ? author.login : undefined
        }));

        commit.authors.map(author => {
          this.logger.veryVerbose.info(`Found author: ${author.username}`);
        });
      })
    );

    return commits;
  }

  public async publish(releaseNotes: string, tag: string) {
    return this.github.publish(releaseNotes, tag);
  }

  public async getLabels(pr: number) {
    return this.github.getLabels(pr);
  }

  public async createStatus(prInfo: IPRInfo) {
    return this.github.createStatus(prInfo);
  }

  public async getSha() {
    return this.github.getSha();
  }

  public async getLatestRelease(): Promise<string> {
    return this.github.getLatestRelease();
  }

  public async getPullRequest(pr: number) {
    return this.github.getPullRequest(pr);
  }

  public async createComment(message: string, pr: number, context = 'default') {
    return this.github.createComment(message, pr, context);
  }

  public async getPullRequests(options?: Partial<GHub.PullsListParams>) {
    return this.github.getPullRequests(options);
  }

  /**
   * Creates new labels on a project that don't already currently exist. If the label
   * is not applicable to the current setup then that label will not be added.
   *
   * @param labels A list of labels to be added to the project
   */
  public async addLabelsToProject(labels: Labels) {
    const oldLabels = await this.github.getProjectLabels();
    const labelsToCreate = Object.entries(labels).filter(
      ([versionLabel, customLabel]) => {
        if (oldLabels && oldLabels.includes(customLabel)) {
          return;
        }

        if (
          versionLabel === Label.release &&
          !this.releaseOptions.onlyPublishWithReleaseLabel
        ) {
          return;
        }

        if (
          versionLabel === Label.skipRelease &&
          this.releaseOptions.onlyPublishWithReleaseLabel
        ) {
          return;
        }

        return true;
      }
    );

    if (!options.dryRun) {
      await Promise.all(
        labelsToCreate.map(async ([versionLabel, customLabel]) => {
          await this.github.createLabel(versionLabel, customLabel);
        })
      );
    }

    const repoMetadata = await this.github.getProject();

    const justLabelNames = labelsToCreate.map(([name]) => name);
    if (justLabelNames.length > 0) {
      const state = options.dryRun ? 'Would have created' : 'Created';
      this.logger.log.log(`${state} labels: ${justLabelNames.join(', ')}`);
    } else {
      const state = options.dryRun ? 'would have been' : 'were';
      this.logger.log.log(
        `No labels ${state} created, they must have already been present on your project.`
      );
    }

    if (!options.dryRun) {
      this.logger.log.log(
        `\nYou can see these, and more at ${repoMetadata.html_url}/labels`
      );
    }
  }

  public async getSemverBump(from: string, to = 'HEAD'): Promise<SEMVER> {
    const commits = await this.getCommits(from, to);
    const labels = commits.map(commit => commit.labels);
    const {
      onlyPublishWithReleaseLabel,
      skipReleaseLabels
    } = this.releaseOptions;
    const options = { onlyPublishWithReleaseLabel, skipReleaseLabels };

    this.logger.verbose.info('Calculating SEMVER bump using:\n', {
      labels,
      versionLabels: this.versionLabels,
      options
    });

    const result = calculateSemVerBump(labels, this.versionLabels, options);

    this.logger.verbose.success('Calculated SEMVER bump:', result);

    return result;
  }

  public async postToSlack(releaseNotes: string, tag: string) {
    if (!this.releaseOptions.slack) {
      throw new Error('Slack url must be set to post a message to slack.');
    }

    const project = await this.github.getProject();

    this.logger.verbose.info('Posting release notes to slack.');

    await postToSlack(releaseNotes, {
      tag,
      owner: this.github.options.owner,
      repo: this.github.options.repo,
      baseUrl: project.html_url,
      slackUrl: this.releaseOptions.slack
    });

    this.logger.verbose.info('Posted release notes to slack.');
  }

  public async calcNextVersion(lastTag: string) {
    const bump = await this.getSemverBump(lastTag);
    return inc(lastTag, bump as ReleaseType);
  }

  private async addLabelsToCommits(commits: ICommit[]) {
    if (!commits) {
      return [];
    }

    const eCommits = normalizeCommits(commits);

    await Promise.all(
      eCommits.map(async commit => {
        if (!commit.pullRequest) {
          commit.labels = [];
        } else {
          commit.labels =
            (await this.getLabels(commit.pullRequest.number)) || [];
        }
      })
    );

    return eCommits;
  }

  private async getPRForRebasedCommits(commits: IExtendedCommit[]) {
    let lastRelease: { published_at: string };

    try {
      lastRelease = await this.github.getLatestReleaseInfo();
    } catch (error) {
      const firstCommit = await this.github.getFirstCommit();
      lastRelease = {
        published_at: await this.github.getCommitDate(firstCommit)
      };
    }

    if (!lastRelease || !lastRelease.published_at) {
      return commits;
    }

    const prsSinceLastRelease = await this.github.searchRepo({
      q: `is:pr is:merged merged:>=${lastRelease.published_at}`
    });
    const pullRequests = await Promise.all(prsSinceLastRelease.items.map(
      async (pr: { number: number }) =>
        this.github.getPullRequest(Number(pr.number))
    ) as GHub.Response<GHub.PullsGetResponse>[]);

    await Promise.all(
      commits.map(async commit => {
        const matchPr = pullRequests.find(
          pr => pr.data.merge_commit_sha === commit.hash
        );

        if (!commit.pullRequest && matchPr) {
          commit.labels = matchPr.data.labels.map(label => label.name) || [];
          commit.pullRequest = {
            number: matchPr.data.number
          };
        }
      })
    );

    return commits;
  }
}
