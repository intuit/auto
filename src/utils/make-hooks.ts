import { AsyncSeriesBailHook, AsyncSeriesHook, SyncHook } from 'tapable';
import { IGitHubReleaseHooks } from '../github-release';
import { ILogParseHooks } from '../log-parse';
import { IAutoHooks } from '../main';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  beforeShipIt: new SyncHook([]),
  onCreateGitHubRelease: new SyncHook(['config']),
  onCreateLogParse: new SyncHook(['gitHubReleaseConfig']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  publish: new AsyncSeriesHook(['version'])
});

export const makeGitHubReleaseHooks = (): IGitHubReleaseHooks => ({
  onCreateLogParse: new SyncHook(['logParser'])
});

export const makeLogHooks = (): ILogParseHooks => ({
  renderChangelogLine: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user'])
});
