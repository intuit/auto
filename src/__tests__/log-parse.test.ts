import { defaultChangelogTitles, defaultLabels } from '../github-release';
import LogParse, {
  filterServiceAccounts,
  IGenerateReleaseNotesOptions,
  normalizeCommits,
  parseJira,
  parsePR,
  parseSquashPR
} from '../log-parse';

import { dummyLog } from '../utils/logger';
import makeCommitFromMsg from './make-commit-from-msg';

describe('createUserLink', () => {
  test('should ', () => {
    const logParse = new LogParse(dummyLog(), {
      owner: '',
      repo: '',
      baseUrl: 'https://github.custom.com/',
      changelogTitles: {},
      versionLabels: defaultLabels
    });
    logParse.loadDefaultHooks();

    expect(
      logParse.createUserLink(
        {
          name: 'none',
          email: undefined,
          username: 'invalid-email-address'
        },
        {
          hash: '1',
          labels: [],
          pullRequest: {
            number: 22
          },
          authorName: 'none',
          authorEmail: 'default@email.com',
          authors: [
            {
              name: 'none',
              email: undefined
            }
          ],
          subject: ''
        }
      )
    ).toBe(undefined);
  });

  test('should find email', () => {
    const logParse = new LogParse(dummyLog(), {
      owner: '',
      repo: '',
      baseUrl: 'https://github.custom.com/',
      changelogTitles: {},
      versionLabels: defaultLabels
    });
    logParse.loadDefaultHooks();

    expect(
      logParse.createUserLink(
        {
          name: 'none',
          email: undefined
        },
        {
          hash: '1',
          labels: [],
          pullRequest: {
            number: 22
          },
          authorName: 'none',
          authorEmail: 'default@email.com',
          authors: [
            {
              name: 'none',
              email: undefined
            }
          ],
          subject: ''
        }
      )
    ).toBe('default@email.com');
  });
});

describe('parsePR', () => {
  test('should do nothing without merge commit', () => {
    const commit = makeCommitFromMsg('Not a merge');
    expect(parsePR(commit)).toEqual(commit);
  });

  test('should parse number base and comment', () => {
    const commit = makeCommitFromMsg(
      'Merge pull request #1234 from Another PR\nComments about the PR'
    );

    expect(parsePR(commit)).toEqual({
      ...commit,
      pullRequest: {
        number: 1234,
        base: 'Another PR'
      },
      subject: 'Comments about the PR'
    });
  });
});

describe('parseSquashPR', () => {
  test('should do nothing squash merge commit', () => {
    const commit = makeCommitFromMsg('Not a squash');
    expect(parseSquashPR(commit)).toEqual(commit);
  });

  test('should parse PR number and comment', () => {
    const commit = makeCommitFromMsg('Some Message (#1234)');
    expect(parseSquashPR(commit)).toEqual({
      ...commit,
      pullRequest: {
        number: 1234
      },
      subject: 'Some Message'
    });
  });
});

describe('filterServiceAccounts', () => {
  test('should not filter the commit', () => {
    const commit = makeCommitFromMsg('foo');
    expect(filterServiceAccounts(commit)).toEqual(commit);
  });

  test('should filter the commit', () => {
    const commit = {
      ...makeCommitFromMsg('foo'),
      authorName: 'pdbf',
      authorEmail: ''
    };

    expect(filterServiceAccounts(commit)).toBeUndefined();
  });
});

describe('jira', () => {
  test('no story', () => {
    const commit = {
      ...makeCommitFromMsg('Add log')
    };

    expect(parseJira(commit)).toEqual(commit);
  });

  test('story found', () => {
    const commit = {
      ...makeCommitFromMsg('Add log'),
      jira: {
        number: ['PLAYA-5052']
      }
    };

    expect(parseJira(makeCommitFromMsg('PLAYA-5052: Add log'))).toEqual(commit);
    expect(parseJira(makeCommitFromMsg('[PLAYA-5052] - Add log'))).toEqual(
      commit
    );
    expect(parseJira(makeCommitFromMsg('[PLAYA-5052] Add log'))).toEqual(
      commit
    );
  });

  test('story found', () => {
    const commit = {
      ...makeCommitFromMsg('Add log'),
      jira: {
        number: ['PLAYA-5052', 'PLAYA-6000']
      }
    };

    expect(
      parseJira(makeCommitFromMsg('PLAYA-5052 PLAYA-6000: Add log'))
    ).toEqual(commit);
    expect(
      parseJira(makeCommitFromMsg('[PLAYA-5052][PLAYA-6000] - Add log'))
    ).toEqual(commit);
    expect(
      parseJira(makeCommitFromMsg('[PLAYA-5052] PLAYA-6000: Add log'))
    ).toEqual(commit);
    expect(
      parseJira(makeCommitFromMsg('PLAYA-5052 [PLAYA-6000] - Add log'))
    ).toEqual(commit);
  });
});

describe('normalizeCommits', () => {
  test('should handle undefined', () => {
    const commits = [makeCommitFromMsg('Filtered', { name: 'pdbf' })];

    expect(normalizeCommits(commits)).toEqual([]);
  });

  test('should do nothing with normal commits', () => {
    const commits = [
      makeCommitFromMsg('First'),
      makeCommitFromMsg('Second'),
      makeCommitFromMsg('Third')
    ];

    expect(normalizeCommits(commits)).toMatchSnapshot();
  });

  test('should use parsing functions to normalize', () => {
    const commits = [
      makeCommitFromMsg('First'),
      makeCommitFromMsg('Second'),
      makeCommitFromMsg('[PLAYA-5052] Add log')
    ];

    expect(normalizeCommits(commits)[2]).toEqual({
      authorEmail: 'adam@dierkens.com',
      authorName: 'Adam Dierkens',
      authors: [
        {
          email: 'adam@dierkens.com',
          name: 'Adam Dierkens'
        }
      ],
      labels: [],
      subject: 'Add log',
      hash: 'foo',
      jira: {
        number: ['PLAYA-5052']
      }
    });
  });
});

const testOptions = (): IGenerateReleaseNotesOptions => ({
  owner: 'foobar',
  repo: 'auto-release',
  baseUrl: 'https://github.custom.com/foobar/auto-release',
  jira: 'https://jira.custom.com/browse',
  versionLabels: defaultLabels,
  changelogTitles: defaultChangelogTitles
});

describe('Hooks', () => {
  test('title', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    const normalized = normalizeCommits([
      makeCommitFromMsg('Some Feature (#1234)')
    ]);

    logParse.hooks.renderChangelogTitle.tap(
      'test',
      (label, changelogTitles) => `:heart: ${changelogTitles[label]} :heart:`
    );
    logParse.loadDefaultHooks();

    expect(await logParse.generateReleaseNotes(normalized)).toMatchSnapshot();
  });

  test('author', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    const normalized = normalizeCommits([
      makeCommitFromMsg('Some Feature (#1234)')
    ]);

    logParse.hooks.renderChangelogAuthor.tap(
      'test',
      (author, commit) => `:heart: ${author.name}/${commit.authorEmail} :heart:`
    );

    logParse.hooks.renderChangelogAuthorLine.tap(
      'test',
      (author, user) => `:shipit: ${author.name} (${user})`
    );
    logParse.loadDefaultHooks();

    expect(await logParse.generateReleaseNotes(normalized)).toMatchSnapshot();
  });
});

describe('generateReleaseNotes', () => {
  test('should create note for PR commits', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    logParse.loadDefaultHooks();
    const normalized = normalizeCommits([
      makeCommitFromMsg('Some Feature (#1234)', { labels: ['minor'] })
    ]);

    expect(await logParse.generateReleaseNotes(normalized)).toMatchSnapshot();
  });

  test('should omit authors with invalid email addresses', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    logParse.loadDefaultHooks();
    const normalized = normalizeCommits([
      makeCommitFromMsg('Some Feature (#1234)', { labels: ['minor'] })
    ]);
    normalized[0].authors[0].username = 'invalid-email-address';

    expect(await logParse.generateReleaseNotes(normalized)).toMatchSnapshot();
  });

  test('should create note for PR commits without labels', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    logParse.loadDefaultHooks();
    const normalized = normalizeCommits([
      makeCommitFromMsg('Some Feature (#1234)')
    ]);

    expect(await logParse.generateReleaseNotes(normalized)).toMatchSnapshot();
  });

  test('should create note for jira commits without labels', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    logParse.loadDefaultHooks();
    const normalized = normalizeCommits([
      makeCommitFromMsg('[PLAYA-5052] - Fix P0')
    ]);

    expect(await logParse.generateReleaseNotes(normalized)).toMatchSnapshot();
  });

  test('should use username if present', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    logParse.loadDefaultHooks();
    const normalized = normalizeCommits([
      makeCommitFromMsg('Some Feature (#1234)', {
        labels: ['minor'],
        username: 'adam'
      })
    ]);

    normalized[0].authors[0].username = 'adam';

    expect(await logParse.generateReleaseNotes(normalized)).toMatchSnapshot();
  });

  test('should create note for JIRA commits', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    logParse.loadDefaultHooks();
    const normalized = normalizeCommits([
      makeCommitFromMsg('[PLAYA-5052] - Some Feature (#12345)', {
        labels: ['major'],
        packages: []
      }),
      makeCommitFromMsg('Some Feature (#1234)', { labels: ['internal'] }),
      makeCommitFromMsg('Third', { labels: ['patch'] })
    ]);

    expect(await logParse.generateReleaseNotes(normalized)).toMatchSnapshot();
  });

  test('should combine pr w/no label and labelled pr', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    logParse.loadDefaultHooks();
    const normalized = normalizeCommits([
      makeCommitFromMsg('Some Feature (#1234)'),
      makeCommitFromMsg('Third', { labels: ['patch'] })
    ]);

    expect(await logParse.generateReleaseNotes(normalized)).toMatchSnapshot();
  });

  test("should use only email if author name doesn't exist", async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    logParse.loadDefaultHooks();
    const commits = normalizeCommits([
      {
        hash: 'foo',
        labels: [],
        authorEmail: 'adam@dierkens.com',
        subject: 'Another Feature (#1234)'
      },
      {
        hash: 'foo',
        labels: [],
        subject: 'One Feature (#1235)'
      }
    ]);

    expect(await logParse.generateReleaseNotes(commits)).toMatchSnapshot();
  });

  test('should include PR-less commits as patches', async () => {
    const logParse = new LogParse(dummyLog(), testOptions());
    logParse.loadDefaultHooks();

    const commits = normalizeCommits([
      {
        hash: '1',
        authorName: 'Adam Dierkens',
        authorEmail: 'adam@dierkens.com',
        subject: 'I was a push to master',
        labels: ['pushToMaster']
      },
      {
        hash: '2',
        authorName: 'Adam Dierkens',
        authorEmail: 'adam@dierkens.com',
        subject: 'First Feature (#1235)',
        labels: ['minor']
      }
    ]);

    expect(await logParse.generateReleaseNotes(commits)).toMatchSnapshot();
  });
});
