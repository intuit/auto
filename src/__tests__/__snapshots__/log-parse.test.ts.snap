// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks author 1`] = `
"#### 🐛  Bug Fix

- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (:heart: Adam Dierkens/adam@dierkens.com :heart:)

#### Authors: 1

:shipit: Adam Dierkens (:heart: Adam Dierkens/adam@dierkens.com :heart:)"
`;

exports[`Hooks title 1`] = `
":heart: 🐛  Bug Fix :heart:
- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)

#### Authors: 1

- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should combine pr w/no label and labelled pr 1`] = `
"#### 🐛  Bug Fix

- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)
- Third  (adam@dierkens.com)

#### Authors: 1

- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should create note for JIRA commits 1`] = `
"#### 💥  Breaking Change

- [PLAYA-5052](https://jira.custom.com/browse/PLAYA-5052): Some Feature [#12345](https://github.custom.com/foobar/auto-release/pull/12345) (adam@dierkens.com)

#### 🐛  Bug Fix

- Third  (adam@dierkens.com)

#### 🏠  Internal

- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)

#### Authors: 1

- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should create note for PR commits 1`] = `
"#### 🚀  Enhancement

- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)

#### Authors: 1

- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should create note for PR commits without labels 1`] = `
"#### 🐛  Bug Fix

- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)

#### Authors: 1

- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should create note for jira commits without labels 1`] = `
"#### 🐛  Bug Fix

- [PLAYA-5052](https://jira.custom.com/browse/PLAYA-5052): Fix P0  (adam@dierkens.com)

#### Authors: 1

- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should include PR-less commits as patches 1`] = `
"#### 🚀  Enhancement

- First Feature [#1235](https://github.custom.com/foobar/auto-release/pull/1235) (adam@dierkens.com)

#### ⚠️  Pushed to master

- I was a push to master  (adam@dierkens.com)

#### Authors: 1

- Adam Dierkens (adam@dierkens.com)"
`;

exports[`generateReleaseNotes should omit authors with invalid email addresses 1`] = `
"#### 🚀  Enhancement

- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234)"
`;

exports[`generateReleaseNotes should use only email if author name doesn't exist 1`] = `
"#### 🐛  Bug Fix

- Another Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) (adam@dierkens.com)
- One Feature [#1235](https://github.custom.com/foobar/auto-release/pull/1235)

#### Authors: 1

- adam@dierkens.com"
`;

exports[`generateReleaseNotes should use username if present 1`] = `
"#### 🚀  Enhancement

- Some Feature [#1234](https://github.custom.com/foobar/auto-release/pull/1234) ([@adam](https://github.custom.com/adam))

#### Authors: 1

- Adam Dierkens ([@adam](https://github.custom.com/adam))"
`;

exports[`normalizeCommits should do nothing with normal commits 1`] = `
Array [
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "subject": "First",
  },
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "subject": "Second",
  },
  Object {
    "authorEmail": "adam@dierkens.com",
    "authorName": "Adam Dierkens",
    "authors": Array [
      Object {
        "email": "adam@dierkens.com",
        "name": "Adam Dierkens",
      },
    ],
    "hash": "foo",
    "labels": Array [],
    "packages": undefined,
    "subject": "Third",
  },
]
`;
